### DISCUSSION TITLE
Junit 5 Unit Tests using Eclipse Junit Container fails with Junit 4 version Error

### DISCUSSION THREAD

**Discussion by jgsuess** [eclipse-tycho/tycho#1282] [Q&A]
I have configured [unit tests within the plugin](https://github.com/eclipse-tycho/tycho/blob/master/RELEASE_NOTES.md#support-resolving-of-junit-classpath-container) in [analogy to the example project ](https://github.com/eclipse-tycho/tycho/blob/master/tycho-its/projects/compiler.junitcontainer/junit4-in-bundle-with-dependencies/pom.xml#L24), but for Junit 5, as below:

```xml
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-surefire-plugin</artifactId>
				<version>3.0.0-M7</version>
				<dependencies>
					<dependency>
						<groupId>org.junit.jupiter</groupId>
						<artifactId>junit-jupiter-engine</artifactId>
						<version>5.8.1</version>
					</dependency>
				</dependencies>
			</plugin>
```

I have taken great care to ensure that the version of the engine I am referencing matches the version provided by the Eclipse container. The test execution fails with the version error below. However, neither the target platform nor the Eclipse Junit container contain 4:10.

```sh
105973 [ERROR] Caused by: org.junit.platform.commons.JUnitException: Unsupported version of junit:junit: 4.10. Please upgrade to version 4.12 or later.
105973 [ERROR] 	at org.junit.vintage.engine.JUnit4VersionCheck.checkSupported(JUnit4VersionCheck.java:49)
105973 [ERROR] 	at org.junit.vintage.engine.JUnit4VersionCheck.checkSupported(JUnit4VersionCheck.java:35)
105973 [ERROR] 	at org.junit.vintage.engine.VintageTestEngine.discover(VintageTestEngine.java:63)
105973 [ERROR] 	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:152)
```

When I remove the engine selection, a version 4.7 is automatically found. It finds no tests, as they are JUnit 5, but it also does not produce the error.

```
Using auto detected provider org.apache.maven.surefire.junit4.JUnit4Provider
```

I am almost certain that this is a configuration problem. So before I try to hunt a bug by distilling a minimal test, I would like to ask here if anyone has seen this before.

**Comment by akurtakov** [comment]
If you want Tycho to do anything you have to use [eclipse-test-plugin](https://wiki.eclipse.org/Tycho/Packaging_Types#eclipse-test-plugin) packaging type and not eclipse-plugin. That would allow for tycho-surefire-plugin to be auto invoked. You're configuring maven-surefire-plugin which has nothing to do with Tycho.

**Reply by jgsuess** [reply]
Thank you, but I think that is not correct. Please read [here](https://github.com/eclipse-tycho/tycho/blob/master/RELEASE_NOTES.md#support-resolving-of-junit-classpath-container) and look at the example integration test linked above. This feature allows the use of regular JUnit tests in plugins, but does not offer OSGi support at that time. I have used it in the past as well.

**Reply by akurtakov** [reply]
My bad. I totally forgot about this capability. Would you please share a minimal reproducer project we can look at?

**Reply by jgsuess** [reply]
I will see if I can obtain the effect from a modified clone of the integration test. I will post here when I get to it.

**Reply by jgsuess** [reply]
See attached. This is modelled on the ITS in master. I can provide a patch if that is preferred.

[junit5 in bundle.zip](https://github.com/eclipse-tycho/tycho/files/9411013/junit5.in.bundle.zip)

**Reply by jgsuess** [reply]
Is there any more advice on this?

**Reply by laeubi** [reply]
@jgsuess if you can transform your example as an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue that would really help. Beside that, have you checked the effective dependencies of your project that they show up? Because actually Tycho is not doing much here (beside compiling the sources) so probably this is a problem in your maven setup or a bug in surefire. In any case enhancing the test-suite with a JUnit5 example would be very useful.
